WIP - Refactoring Invoke-Pester and early configuration-validation

fflaten opened the pull request:
## PR Summary
Playing around with multiple refactoring of `Invoke-Pester` to make it more readable + improving configuration validation. Ex.
- [ ] Extract parameter set migration for Simple and Legacy-set
- [ ] Call TestResult-export through plugin End-step
- [ ] Move CodeCoverage processing to plugin-steps Start, RunEnd and End
- [ ] Remove legacy comment help? Ex. hashtable-format for -Script, CodeCoverage etc.
- [ ] Validate configuration early to avoid running a long run just to loose it at the end when it throws due to invalid format for coverage, testresult etc. (ex. #2304)
  - Throwing in Start plugin-step creates an ugly and duplicate error atm. Likely due to being thrown through `Assert-Success`. Might be better to use own function like `Resolve-OutputConfiguration`

## PR Checklist

- [ ] PR has meaningful title
- [ ] Summary describes changes
- [ ] PR is ready to be merged
  - If not, use the arrow next to `Create Pull Request` to mark it as a draft. PR can be marked `Ready for review` when it's ready.
- [ ] Tests are added/update *(if required)*
- [ ] Documentation is updated/added *(if required)*

fflaten commented:
>   * Throwing in Start plugin-step creates an ugly and duplicate error atm. Likely due to being thrown through `Assert-Success`. Might be better to use own function like `Resolve-OutputConfiguration`

The duplicate was because `Assert-Success` printed the error while also throwing it. Suggest removing the host message.

Thoughts on where to validate config, especially string-options? Suggestions:
1. Start plugin-step. Removed the duplicate message, but still a bit verbose due to `Assert-Success`
![image](https://user-images.githubusercontent.com/3436158/223556006-9af85e5b-fb88-4105-88a3-cbef61921ba4.png)

2. Validation-function, ex. `Resolve-OutputConfiguration`
![image](https://user-images.githubusercontent.com/3436158/223556147-4154acc6-65ac-4d7d-b527-7a3e6f02f488.png)

3. Validate in assembly for StringOption. Would not support public plugins integrating, but those plugins should probably not replace built-in options.
![image](https://user-images.githubusercontent.com/3436158/223569901-2e345578-516e-4bed-bc87-eb9c6944e1d8.png)

5. A new `ConfigurePlugin/SetupPlugin` step invoked without `Invoke-PluginStep` or a different branch to simplify error. Maybe invoked before `Find-File` call which can be slow. This might fit better with public plugins.

fflaten commented on src/Pester.Runtime.ps1:
```diff
@@ -1289,21 +1300,20 @@ function Assert-Success {
     foreach ($r in $InvocationResult) {
+        $rc++
         $ec = 0
         if ($null -ne $r.ErrorRecord -and $r.ErrorRecord.Length -gt 0) {
-            $err += "Result $($rc++):"
             $anyFailed = $true
             foreach ($e in $r.ErrorRecord) {
-                $err += "Error $($ec++):"
+                $err += "$([Environment]::NewLine)Result $rc - Error $((++$ec)):"
```
Previous message didn't print `$ec` and `$rc` (only incremented them) due to missing parantheses.

fflaten commented:
/azp run

nohwnd commented on src/Pester.Runtime.ps1:
```diff
@@ -1216,7 +1227,7 @@ function Invoke-PluginStep {
     # only do this if there is more than 1, to avoid the "expensive" -like check and reverse
     $isEndStep = 1 -lt $pluginsWithGivenStep.Count -and $Step -like "*End"
-    if (-not $isEndStep) {
+    if ($isEndStep) {
```
So this did not reverse the steps before, did you see any bugs linked to that?

fflaten replied:
No bugs I'm aware of. Just not behaving like comment says. We reverse on TearDown-steps as expected, but this currently reverses on everything else than *End.

Not a critical part of the PR, just testing it out while in draft.

fflaten commented:
> > * Throwing in Start plugin-step creates an ugly and duplicate error atm. Likely due to being thrown through `Assert-Success`. Might be better to use own function like `Resolve-OutputConfiguration`
> 
> The duplicate was because `Assert-Success` printed the error while also throwing it. Suggest removing the host message.
> 
> Thoughts on where to validate config, especially string-options? Options:
> 
> 1. Start plugin-step. Removed the duplicate message, but still a bit verbose due to `Assert-Success`
>    ![image](https://user-images.githubusercontent.com/3436158/223556006-9af85e5b-fb88-4105-88a3-cbef61921ba4.png)
> 2. Validation-function, ex. `Resolve-OutputConfiguration`
>    ![image](https://user-images.githubusercontent.com/3436158/223556147-4154acc6-65ac-4d7d-b527-7a3e6f02f488.png)
> 3. Validate in assembly for StringOption. Would block third-party plugins from integrating with built-in options like `TestResult.OutputFormat`, but do we want to support that?
>    ![image](https://user-images.githubusercontent.com/3436158/223569901-2e345578-516e-4bed-bc87-eb9c6944e1d8.png)
>    ![image](https://user-images.githubusercontent.com/3436158/223579128-fd3a2ad1-3c08-49fd-87b8-4424c32be1a9.png)
> 4. A new `ConfigurePlugin/SetupPlugin` step invoked without `Invoke-PluginStep` or a different branch to simplify error. Maybe invoked before `Find-File` call which can be slow. This might also benefit public plugin API.

Thoughts on this @nohwnd ? Use validation-functions for now and consider a plugin-step later as part of making plugins public?

fflaten commented:
Rebased and updated with validation-functions for now. Can revisit later when looking into public plugins API.

Also cleaned up the error message for invalid StringOption. It now list option and supported values just like CodeCoverage did.
![image](https://github.com/pester/Pester/assets/3436158/e161baa6-7252-4683-b1d2-3b6156769bc3)
![image](https://github.com/pester/Pester/assets/3436158/285ffe9c-43e2-4df7-bf5d-18586acb3ade)

fflaten commented on src/Pester.Runtime.ps1:
```diff
@@ -1292,21 +1303,20 @@ function Assert-Success {
 
     if ($anyFailed) {
-        $Message = $Message + ":`n$err"
-        Write-PesterHostMessage -ForegroundColor Red $Message
```
Revert this or good? Not validating configuration in plugin-step now, so PR is not affected by the duplicate error from `Assert-Success`. Not sure if it's useful somewhere else

nohwnd replied:
good.

nohwnd commented:
@fflaten  Approved both this and your test output refactoring PRs, they seem like they might run into each other so I was not sure in which order you want them merged, please merge any way you like :)

fflaten commented:
@nohwnd Thanks! Let's start with this. It will also cause a conflict in the last PR for container data. Can fix both later today. 🙂

fflaten commented:
Btw. Don't have merge rights, so up to you :)

nohwnd commented:
@fflaten  you do now.
